more coroutines

havardpe opened the pull request:
@vekterli please review

Do not merge just yet. I am seeing local valgrind errors complaining about leaked tls storage or something.

havardpe committed:
more coroutines

havardpe committed:
Merge branch 'master' into havardpe/more-coroutines

vekterli commented on vespalib/src/vespa/vespalib/coro/lazy.h:
```diff
@@ -5,10 +5,23 @@
+ * The coroutine is lazy (will suspend in initial_suspend) and
+ * destroyed form the outside (will suspend in final_suspend). Waiting
+ * for a Lazy<T> using co_await will use symmetric transfer to suspend
+ * the waiting coroutine and resuming this one. The waiting coroutine
```
```suggestion
 * the waiting coroutine and resume this one. The waiting coroutine
```

vekterli commented on vespalib/src/vespa/vespalib/coro/lazy.h:
```diff
@@ -5,10 +5,23 @@
+ * for a Lazy<T> using co_await will use symmetric transfer to suspend
+ * the waiting coroutine and resuming this one. The waiting coroutine
+ * is registered as a continuation and will be resumed again once the
+ * result is avalable (also using symmetric transfer). The result is
```
```suggestion
 * result is available (also using symmetric transfer). The result is
```

vekterli reviewed (approved):
✨🦄👍

havardpe committed:
add suppression and make safer

- add new valgrind suppression variant for leaked thread stack cache
- hold lock while editing list of threads
- use jthread for join in destructor

havardpe committed:
Apply suggestions from code review

typos

Co-authored-by: Tor Brede Vekterli <vekterli@yahooinc.com>

havardpe commented:
@vekterli PTAL

vekterli reviewed (approved):
:+1:
